### PR TITLE
Add zig tree-sitter support

### PR DESCRIPTION
**Pull request by voroskoi**
Highlighting works, indenting not so much, but I guess it is more than nothing.
Queries are from https://github.com/Himujjal/tree-sitter-zig

### CODE REVIEW

**Comment by kirawi** [comment]
`indents.scm` and `folds.scm` are not supported. For indents you'll have to define a `indents.toml` instead, which looks like this: https://github.com/helix-editor/helix/blob/master/runtime/queries/rust/indents.toml

File: runtime/queries/zig/highlights.scm [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  function: (identifier) @function)
+
+(build_in_call_expr
+  function: (identifier) @function.builtin
+)
+
+;; other identifiers
+(type_identifier) @type
+(primitive_type) @type.builtin
+(field_identifier) @field
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:31]: `field` is not a scope we use, it should be `property`.

File: runtime/queries/zig/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+(string_literal) @string
+(multiline_string_literal) @string 
+
+(escape_sequence) @constant.builtin
+
+(label_identifier) @label
+
+(call_modifier) @keyword ; async
+
+(binary_operator) @keyword.operator
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:55]: Change this to just `operator`.

File: runtime/queries/zig/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+(field_identifier) @field
+
+(line_comment) @comment
+(doc_comment) @comment
+
+(char_literal) @number
+(integer_literal) @number
+(float_literal) @number
+
+(boolean_literal) @boolean
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:40]: This should be `constant.builtin`.

File: runtime/queries/zig/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+] @keyword
+
+"fn" @keyword.function
+
+[
+  "continue"
+  "else"
+  "if"
+  "switch"
+] @conditional
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:117]: This should be `keyword.control`. I'm not sure whether it would make sense to change it to `keyword.control.conditional`, I'm concerned about having unstandardized scopes @archseer .

File: runtime/queries/zig/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+[
+  "for"
+  "while"
+] @repeat
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:122]: This should also be `keyword.control`.

**Review by kirawi — CHANGES_REQUESTED** [review]

File: runtime/queries/zig/highlights.scm (hunk 1 above)
Comment (voroskoi) [reply runtime/queries/zig/highlights.scm:31]: How do I know which scopes are available? I have no idea where these come from.

Comment (kirawi) [reply runtime/queries/zig/highlights.scm:31]: Ah, it's listed here: https://docs.helix-editor.com/themes.html

**Comment by voroskoi** [comment]
Thank You for the heads up, I have cleaned up things a bit. I have switched to a different grammar repo, but nvim-treesitter also use this one, so it is probably more up to date. Still have to figure out the indenting, does not work for me.

File: runtime/queries/zig/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+
+; functionn decl
+(FnProto
+  (IDENTIFIER) @function
+  (SuffixExpr (IDENTIFIER) @type)?
+  ("!")? @function.macro
+)
+
+(ParamDecl 
+  (ParamType (SuffixExpr (IDENTIFIER) @parameter))
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:101]: should be `@variable.parameter`

File: runtime/queries/zig/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+)
+
+(ParamDecl 
+  (IDENTIFIER) @parameter
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:105]: Should also be `@variable.parameter`

File: runtime/queries/zig/highlights.scm [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    "."
+    .
+    (IDENTIFIER) @constant
+  )
+)
+
+(BUILTINIDENTIFIER) @function.builtin
+
+((BUILTINIDENTIFIER) @include
+  (#any-of? @include "@import" "@cImport"))
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:124]: These should probably just be `@keyword`. I wonder if `#any-of?` is supported.

File: runtime/queries/zig/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+(INTEGER) @number
+
+(FLOAT) @number
+
+[
+  (STRINGLITERAL)
+  (STRINGLITERALSINGLE)
+] @string
+
+(CHAR_LITERAL) @character
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:135]: Should be `@string`

File: runtime/queries/zig/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+  "undefined"
+  "unreachable"
+  "null"
+] @constant.builtin
+
+[
+  "else"
+  "if"
+  "switch"
+] @keyword.control
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:163]: These should be collapsed into one block.

File: runtime/queries/zig/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+  "test"
+  "pub"
+  "usingnamespace"
+] @keyword
+
+[
+  "return"
+  "break"
+  "continue"
+] @keyword
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:210]: These should be `@keyword.control`

File: runtime/queries/zig/highlights.scm [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+  "defer"
+  "errdefer"
+  "async"
+  "nosuspend"
+  "await"
+  "suspend"
+  "resume"
+  "export"
+  "extern"
+] @function.macro
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:223]: I'm not familiar with Zig, but are these related macros? They sound like a mix of `@keyword` and `@keyword.control`

File: runtime/queries/zig/highlights.scm [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+  ";"
+  "."
+  ","
+  ":"
+] @punctuation.delimiter
+
+[
+  ".."
+  "..."
+] @punctuation
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:265]: These should also be collapsed into one block.

File: runtime/queries/zig/highlights.scm [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+  "threadlocal"
+  "fn"
+] @keyword.directive
+
+[
+  "test"
+  "pub"
+  "usingnamespace"
+] @keyword
+
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:205]: I would suggest collapsing all `@keyword` matches into one block as well.

File: runtime/queries/zig/injections.scm [hunk 15]
```diff
@@ -0,0 +1,5 @@
+[
+  (container_doc_comment)
+  (doc_comment)
+  (line_comment)
+] @comment
```
Comment (kirawi) [inline runtime/queries/zig/injections.scm:5]: I'm not sure that this would work, since you also have to specify the injection language like here: https://github.com/helix-editor/helix/blob/master/runtime/queries/rust/injections.scm

File: runtime/queries/zig/highlights.scm [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+  (MultiplyOp)
+  (PrefixOp)
+  "*"
+  "**"
+  "->"
+  "=>"
+  ".?"
+  ".*"
+  "="
+] @operator
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:253]: These should also be grouped up with the other `@operator` blocks into one `@operator` block.

File: runtime/queries/zig/highlights.scm [hunk 17]
```diff
@@ -0,0 +1,277 @@
+[
+  (container_doc_comment)
+  (doc_comment)
+  (line_comment)
+] @comment
+
+(IDENTIFIER) @variable
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:7]: Does highlighting really work? I was under the assumption that it would highlighting everything as a variable at this point (since I don't think you can override them). I might be wrong since `locals.scm` is involved.

**Review by kirawi — CHANGES_REQUESTED** [review]
You're getting there!

File: runtime/queries/zig/injections.scm (hunk 15 above)
Comment (voroskoi) [reply runtime/queries/zig/injections.scm:5]: I just remove them for now, does not really matter.

File: runtime/queries/zig/highlights.scm (hunk 17 above)
Comment (voroskoi) [reply runtime/queries/zig/highlights.scm:7]: No, it does not, You are right.

File: runtime/queries/zig/highlights.scm (hunk 8 above)
Comment (voroskoi) [reply runtime/queries/zig/highlights.scm:124]: Removed this one, does not matter.

**Comment by voroskoi** [comment]
Ok, I think the highlights part is fine for initial support. Time to figure out the indenting part.

File: runtime/queries/zig/highlights.scm [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+
+(SwitchItem 
+  (SuffixExpr
+    "."
+    .
+    (IDENTIFIER) @constant
+  )
+)
+
+(BUILTINIDENTIFIER) @function.builtin
```
Comment (kirawi) [inline runtime/queries/zig/highlights.scm:119]: This should also be in the `@function.builtin` block.

**Comment by voroskoi** [comment]
I have removed the `indents.toml` as I can not figure out why it does not do anything.

**Review by kirawi — CHANGES_REQUESTED** [review]

File: runtime/queries/zig/highlights.scm (hunk 18 above)
Comment (voroskoi) [reply runtime/queries/zig/highlights.scm]: Done.

**Comment by kirawi** [comment]
It seems like the highlighting is a little incorrect:
Capture

Function calls are being highlighted as variables.

**Comment by archseer** [comment]
> I have removed the `indents.toml` as I can not figure out why it does not do anything.

Can you show me the file (it's gone from git history)? I might be able to help there.

**Comment by archseer** [comment]
Nice work! I'll do some testing on indents but I think it's good for merge 👍
